Rename surface forcing tice

jm-c opened the pull request:
## What changes does this PR introduce?
- rename array "surfaceForcingTice" to "adjustColdSST_diag" to better reflect
  what it is used for ; 
- also remove most of the TAF dir storage which are not necessary
  once adjustColdSST_diag is properly initialised.

## What is the current behaviour? 
  name "surfaceForcingTice" was confusing and description in FFIELDS.h
  even more (not really a forcing and has little to do with seaice
  formation and even less with seaice melting).

## What is the new behaviour 
Change the name and description to better reflect what it is,
  a diagnostic for SST adjustment.

## Does this PR introduce a breaking change? 
no

## Other information:
The  only TAF storage dir left is related to KPP, and has been put within #idef ALLOW_KPP / #endif ;
While it is not very clear why this adjustment diag is affecting KPP (because it's effects have already
been added to the solution and stratification), if this needs to be changed, it would be in an other PR.

## Suggested addition to `tag-index`
o model/src:
 - rename array "surfaceForcingTice" to  "adjustColdSST_diag" and  change 
   the description to better reflect what it is, a diagnostic for SST adjustment.
- remove most of the TAF storage directive for "adjustColdSST_diag" that were
  not necessary once it's properly initialized.

jm-c commented:
One more thing to change: shift StoreDynVars2D slot number after removing surfaceforcingTice
in autodiff_store.F/_restore.F and reduce NDV2D by 1 in AUTODIFF.h

mjlosch commented:
Hi @jm-c ,
I don't remember the history of this (maybe @menemenlis does), but I think this extra forcing is necessary for the counter gradient transport of KPP, because is the implied surface forcing that would move theta back to freezing temperatures. For this reason, I don't think that the new name is any better than the old one (I was never bothered by the old name because of the analogy with surfaceForcingT), but I have no strong opinion about this. Is this ever used by anyone anymore?
The implementation (of surfaceForcingTice) is inconsistent, probably to make sure that theta really is at freezing. I guess one could add this "forcing" to surfaceForcingT explicitly and not reset theta in this routine, but that would probably lead to other problems, wouldn't it?

Everything else looks good to me, especially removing unnecessary storage. So I approve of this.

jm-c commented:
@mjlosch I would like to make some comments regarding confusion around surfaceForcingTice:
1) I agree, allowFreezing is not much used, and it's why I would prefer not to 
  change the current behavior. However, I would like to avoid confusion among users
  about when "allowFreezing=T" should be used and what to expect in array
  surfaceForcingTice (new name: adjustColdSST_diag) and associated diagnostic "oceFreez".
 a) When I use a seaice pkg, I am tempted to use "allowFreezing=T" (to allow seawater
  to freeze and form seaice) otherwise I have the impression seaice will only be able 
  to melt or to form from accumulation of snow ?
  This PR does not address this confusing parameter name (and description), but this
  would involve changes in namelist and model parameter input file "data", so leave 
  this for an other PR.
 b) from the description of surfaceForcingTice in FFIELDS.h:
C            - equivalent Temperature flux in the top level that corresponds
C              to the melting or freezing of sea-ice.
C              Note that the surface level temperature is modified
C              directly by the sea-ice model in order to maintain
C              water temperature under sea-ice at the freezing point. 
       and more related to KPP ...
 it's clear from this description that, when I use a seaice pkg, I expect to find 
 in "surfaceForcingTice" all the latent heat-flux associated with melting (seaice) 
 and freezing (sea water). This is very confusing.

2) what is done in freeze_surface.F is an adjustment to SST (or correction to SST):
    line 59:
   theta(i,j,k,bi,bj) = Tfreezing
  And in order to provide a closed heat budget, the equivalent heating that this 
  adjustment implies is stored into array: 
   surfaceForcingTice (new name: adjustColdSST_diag).
  Apart from KPP (will come back to it), surfaceForcingTice is just used for 
  diagnostic and nothing else.
  This is the reason why I suggest to rename "surfaceForcingTice" and I suggested
  this new name.
  Now, I agree with your last point, this hack for too cold SST when no seaice
  pkg is used could have been coded differently, more like a surface forcing
  additional component (next to surfaceForcingT) and added in APPLY_FORCING_T
  but my point is that it's not currently what the model does.

3) The problem with current surfaceForcingTice in KPP:
   the model state and the stratification that KPP sees already account for 
   this SST correction, but instead of providing to KPP a surface buoyancy 
   forcing that is consistent with what the model uses, we change it to account 
   for this SST adjustment that has already been applied.
   Now the way it's currently done might have some advantages, but it is not 
   a very natural way to deal with such a simple correction.
   And regarding the non-local term, it's also not very natural to consider
   a different surface forcing (to redistribute at depth) from what the model see.
   For these reasons, I added a comment/note in FFIELDS.h
C               2) this diagnostic is currently used by KPP package (kpp_calc.F
C              and kpp_transport_t.F) although it is not very clear it should.

   Now I think that a better place to call S/R FREEZE_SURFACE would be where
   other adjustment/correction/filter to theta are done, i.e., 
   in tracers_correction_step.F
   If it was the case, we would probably not try to use surfaceForcingTice
   in KPP calculation and non-local term.

jm-c commented:
@menemenlis do you have comments/suggestions ?

christophernhill commented:
@menemenlis can we get your input - please?

jm-c commented:
Going to merge soon.
